### DISCUSSION TITLE
How to define a user-defined TYPE (new feature spotted today)?

### DISCUSSION THREAD

**Discussion by PowerAppsDarren** [microsoft/Power-Fx#2722] [General]
How to define a user-defined TYPE (new feature spotted today)?

![image](https://github.com/user-attachments/assets/b65c8b9b-fff9-4641-9b91-7ef145fb02d2)

![image](https://github.com/user-attachments/assets/79f89fe7-8efa-44c2-b4d7-6efa2f6128f6)

From: https://github.com/microsoft/Power-Fx/blob/main/releasenotes/releasenotes-1.3.0-rc.md

•	https://github.com/microsoft/Power-Fx/pull/2569

 
![image](https://github.com/user-attachments/assets/4dfdbfe9-255b-4c87-876d-1d4fcd5701c5)

**Comment by anderson-joyle** [comment]
You can find some good examples in the test files:
https://github.com/microsoft/Power-Fx/blob/main/src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/IsType_UO.txt
https://github.com/microsoft/Power-Fx/blob/main/src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt

Some of the test cases:
`AsType(ParseJSON("{""foo"": true, ""bar"": 1.1}"), Type({foo: Boolean, bar: Number}))`
`AsType(ParseJSON("{""Name"": ""SpongeBob"", ""Age"": 1}"), Type({Name: Text, Age: Number, Aquatic: Boolean})).Name`
`IsType(ParseJSON("{""foo"": true, ""bar"": 1.1}"), Type({foo: Boolean, bar: Number}))`
`AsType(ParseJSON("[{""a"": ""Hello"", ""b"": ""2012-01-02""}, {""a"": ""Hi"", ""b"": ""2012-01-03""}]"), Type([{a: Text, b: Date}]))`

**Reply by PowerAppsDarren** [reply]
So, when using these with user-defined functions, do we pick, "UntypedObject"?

---
![image](https://github.com/user-attachments/assets/68da2931-263d-4a1c-895d-38dada8a8380)
---
![image](https://github.com/user-attachments/assets/5aa8142c-3b1d-45dc-810c-b640622d61c7)
---
![image](https://github.com/user-attachments/assets/ab64718c-7e0f-4d84-abf6-ab40ab3909b3)
---
A working example of a user-defined function using a user-defined type in App.Formulas would be very helpful! 😀

...thank you!

**Reply by PowerAppsDarren** [reply]
These are the types we get in Power Apps studio right now: 

![image](https://github.com/user-attachments/assets/0f40ed85-73d4-473c-8789-449a1621cc34)

**Comment by PowerAppsDarren** [comment]
Here we go. This works great: 
  
```
fxMyPerson := Type({FirstName: Text, LastName: Text});

fxPeople := Type([fxMyPerson]);
```
